### PR TITLE
Load initial assignment and dynamically watch/apply keyspace groups' membership/distribution change

### PR DESCRIPTION
**Pull request by binshi-bing**
### What problem does this PR solve?

Issue Number: Ref #6232 

### What is changed and how does it work?

```commit-message
Load initial keyspace group assignment.
Dynamically watch/apply keyspace groups' membership/distribution change.
```

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Code changes

- Has the configuration change
- Has HTTP API interfaces changed (Don't forget to [add the declarative for the new API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))
- Has persistent data change

Side effects

- Possible performance regression
- Increased code complexity
- Breaking backward compatibility

Related changes

- PR to update [`pingcap/docs`](https://github.com/pingcap/docs)/[`pingcap/docs-cn`](https://github.com/pingcap/docs-cn):
- PR to update [`pingcap/tiup`](https://github.com/pingcap/tiup):
- Need to cherry-pick to the release branch

### Release note

```release-note
None.
```

### CODE REVIEW

**Comment by lhy1024** [comment]
cc @lhy1024

File: pkg/tso/keyspace_group_manager.go [hunk 1]
```diff
@@ -94,63 +132,398 @@ func NewKeyspaceGroupManager(
...
+	kgs := make([]*endpoint.KeyspaceGroup, 0, len(resp.Kvs))
+	for _, item := range resp.Kvs {
+		kg := &endpoint.KeyspaceGroup{}
+		if err = json.Unmarshal(item.Value, kg); err != nil {
+			return 0, nil, err
+		}
+		kgs = append(kgs, kg)
+	}
+
+	return resp.Header.Revision, kgs, nil
```
Comment (rleungx) [inline pkg/tso/keyspace_group_manager.go:316]: How about checking Header to make sure it won't be nil?

File: pkg/tso/keyspace_group_manager.go [hunk 2]
```diff
@@ -16,54 +16,87 @@ package tso
...
 )
 
-// primaryElectionSuffix is the suffix of the key for keyspace group primary election
-const primaryElectionSuffix = "primary"
+const (
+	// primaryElectionSuffix is the suffix of the key for keyspace group primary election
+	primaryElectionSuffix = "primary"
+	// defaultLoadKeyspaceGroupsTimeout is the default timeout for loading the initial
+	// keyspace group assignment
+	defaultLoadKeyspaceGroupsTimeout   = 3 * time.Minute
```
Comment (rleungx) [inline pkg/tso/keyspace_group_manager.go:48]: Is 3 min too long?

**Review by rleungx — COMMENTED** [review]
The rest LGTM

File: pkg/tso/keyspace_group_manager.go [hunk 3]
```diff
@@ -73,17 +106,25 @@ type KeyspaceGroupManager struct {
...
 	// Note: The {group} is 5 digits integer with leading zeros.
 	tsoSvcRootPath string
+	// legacySvcStorage is storage with legacySvcRootPath.
+	legacySvcStorage *endpoint.StorageEndpoint
+	// tsoSvcStorage is storage with tsoSvcRootPath.
+	tsoSvcStorage *endpoint.StorageEndpoint
 	// cfg is the TSO config
 	cfg           ServiceConfig
 	maxResetTSGap func() time.Duration
+	// loadKeyspaceGroupsTimeout is the timeout for loading the initial keyspace group assignment
```
Comment (lhy1024) [inline pkg/tso/keyspace_group_manager.go:116]:
```suggestion
	// loadKeyspaceGroupsTimeout is the timeout for loading the initial keyspace group assignment.
```

File: pkg/tso/keyspace_group_manager.go [hunk 4]
```diff
@@ -94,63 +135,413 @@ func NewKeyspaceGroupManager(
...
+					zap.Int64("compact-revision", wresp.CompactRevision))
+				revision = wresp.CompactRevision
+				break
+			}
+			if wresp.Canceled {
+				log.Error("watcher is canceled",
+					zap.Int64("required-revision", revision),
+					errs.ZapError(errs.ErrEtcdWatcherCancel, wresp.Err()))
+				return revision, wresp.Err()
+			}
```
Comment (lhy1024) [inline pkg/tso/keyspace_group_manager.go:382]: Do we need to solve other errors?

File: pkg/tso/keyspace_group_manager.go [hunk 5]
```diff
@@ -94,63 +135,413 @@ func NewKeyspaceGroupManager(
...
+		select {
+		case <-kgm.ctx.Done():
+			return revision, nil
+		default:
+		}
+	}
+}
+
+func (kgm *KeyspaceGroupManager) isAssignedToMe(group *endpoint.KeyspaceGroup) bool {
+	var assignedToMe = false
```
Comment (lhy1024) [inline pkg/tso/keyspace_group_manager.go:417]: It seems to be unnecessary. We can return true directly when found the member.

File: pkg/tso/keyspace_group_manager.go [hunk 6]
```diff
@@ -94,63 +135,413 @@ func NewKeyspaceGroupManager(
...
+	kgm.mu.Lock()
+	defer kgm.mu.Unlock()
+
+	select {
+	case <-ctx.Done():
+		return
+	default:
+	}
+
+	if group.ID >= uint32(len(kgm.ams)) {
```
Comment (lhy1024) [inline pkg/tso/keyspace_group_manager.go:439]: Is it better to use map as `kgm.ams`? only a idea.

File: pkg/tso/keyspace_group_manager.go [hunk 7]
```diff
@@ -94,63 +135,413 @@ func NewKeyspaceGroupManager(
...
+// deleteKeyspaceGroup deletes the given keyspace group.
+func (kgm *KeyspaceGroupManager) deleteKeyspaceGroup(id uint32) {
+	kgm.mu.Lock()
+	defer kgm.mu.Unlock()
+
+	select {
+	case <-kgm.ctx.Done():
+		return
+	default:
+	}
```
Comment (lhy1024) [inline pkg/tso/keyspace_group_manager.go:495]: Is it necessary?

**Review by lhy1024 — COMMENTED** [review]
the rest LGTM

File: pkg/tso/keyspace_group_manager.go [hunk 8]
```diff
@@ -94,63 +130,392 @@ func NewKeyspaceGroupManager(
...
+		kv.NewEtcdKVBase(kgm.etcdClient, kgm.tsoSvcRootPath), nil)
+	return kgm
 }
 
 // Initialize this KeyspaceGroupManager
-func (kgm *KeyspaceGroupManager) Initialize() {
-	// TODO: dynamically load keyspace group assignment from the persistent storage and add watch for the assignment change
-	kgm.initDefaultKeyspaceGroup()
+func (kgm *KeyspaceGroupManager) Initialize(loadFromStorage bool) error {
+	kgm.legacySvcStorage = endpoint.NewStorageEndpoint(
```
Comment (JmPotato) [inline pkg/tso/keyspace_group_manager.go:156]: Why create the `legacySvcStorage` twice here? It has already been initialized in `NewKeyspaceGroupManager` function.

File: pkg/tso/keyspace_group_manager.go [hunk 9]
```diff
@@ -94,63 +130,392 @@ func NewKeyspaceGroupManager(
...
+		kgm.updateKeyspaceGroup(group)
+		return nil
+	}
+
+	// Load the initial keyspace group assignment from storage with time limit
+	done := make(chan struct{}, 1)
+	ctx, cancel := context.WithCancel(kgm.ctx)
+	go kgm.checkInitProgress(ctx, cancel, done)
+	watchStartRevision, err := kgm.initAssignment(ctx)
+	done <- struct{}{}
```
Comment (JmPotato) [inline pkg/tso/keyspace_group_manager.go:178]: Does a `context.WithTimeout` could replace `checkInitProgress`?

**Review by rleungx — APPROVED** [review]

**Review by JmPotato — APPROVED** [review]
Basically LGTM. However, should we consider decoupling `keyspace_group` from the TSO package? This term could potentially serve as a universal reference when developing other microservices such as the resource manager.

**Pull request merged**